### PR TITLE
Migrate socket module from ballerina-lang including history

### PR DESCRIPTION
**Pull request by BuddhiWathsala**
## Purpose
$subject

## Approach
- Move the code base of the socket module out from the ballerina-lang repository with history
- Introduce two sub-modules
    - socket-utils - Java util code used in the socket module
    - socket-generator - Ballerina code that generates the final artifacts
- Change the Gradle configurations

## Related Issues
https://github.com/ballerina-platform/ballerina-lang/issues/21313

## Test environment
- Java 8
- Ballerina 1.2.0
- MacOS
 
## Learning
Configuration changes done referring to these repositories [1, 2]. For more details about the migration refer to this mail thread [3].

[1] https://github.com/ballerinax/docker
[2] https://github.com/ballerina-platform/module-mongodb
[3] https://groups.google.com/forum/#!topic/ballerina-dev/RssHIxkR1XA

### CODE REVIEW

File: .github/workflows/gradle.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    runs-on: windows-latest
+
+    steps:
+      - uses: actions/checkout@v1
+      - name: Set up JDK 1.8
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+      - name: Build with Gradle
+        run: ./gradlew.bat build
```
Comment (daneshk) [inline .github/workflows/gradle.yml:30]:
```suggestion
        run: ./gradlew.bat build

```

File: build.gradle [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+ext.ballerinaLangVersion = '1.3.0-SNAPSHOT'
+ext.puppycrawlCheckstyleVersion = "8.18"
+ext.testngVersion = "6.14.3"
+ext.slf4jVersion = "1.7.30"
+
+allprojects {
+    group = 'org.ballerinalang'
+    version = '1.3.0-SNAPSHOT'
```
Comment (daneshk) [inline build.gradle:32]:
```suggestion
    version = ballerinaLangVersion
```

Comment (daneshk) [reply build.gradle:32]: Shall we use the `ballerinaLangVersion` variable defined earlier

File: socket-ballerina/build.gradle [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+import org.apache.tools.ant.taskdefs.condition.Os
+
+apply from: "$rootDir/gradle/mvnPublisher.gradle"
+
+description = 'Ballerinax - Socket Ballerina Generator'
```
Comment (daneshk) [inline socket-ballerina/build.gradle:22]:
```suggestion
description = 'Ballerina - Socket Generator'
```

File: socket-native/build.gradle [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * limitations under the License.
+ *
+ */
+
+plugins {
+    id 'java'
+}
+apply from: "$rootDir/gradle/javaProject.gradle"
+
+description = 'Ballerinax - Socket Java Utils'
```
Comment (daneshk) [inline socket-native/build.gradle:23]:
```suggestion
description = 'Ballerina - Socket Java Utils'
```

File: gradle/wrapper/gradle-wrapper.properties [hunk 5]
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.2-bin.zip
```
Comment (wggihan) [inline gradle/wrapper/gradle-wrapper.properties:3]: Shall we test with the latest gradle version?

File: socket-ballerina/build.gradle [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    socketBallerinaJar
+}
+
+dependencies {
+    balTools ("org.ballerinalang:jballerina-tools:${ballerinaLangVersion}") {
+        transitive = false
+    }
+    compile project(':socket-native')
+}
+
```
Comment (wggihan) [inline socket-ballerina/build.gradle:36]: Let's remove these additional empty lines.

File: socket-ballerina/build.gradle [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            // Copying the jar from ballerina build
+            from "$project.projectDir/target/caches/jar_cache/ballerina/socket/ballerina-socket-.jar"
+            into "$project.projectDir/target/caches/jar_cache/ballerina/socket/"
+            rename { String fileName ->
+                fileName.replace("ballerina-socket-.jar", artifactSocketBallerinaJar.getName())
+            }
+        }
+        // Copy the BIR to the 0.0.0 folder within it
+        ant.move file: "${artifactSocketBallerinaBirDir}/ballerina/socket/socket.bir",
+                todir: "${artifactSocketBallerinaBirDir}/ballerina/socket/0.0.0/"
```
Comment (wggihan) [inline socket-ballerina/build.gradle:90]: What is this 0.0.0? Is that something internal or socket module version?

Comment (BuddhiWathsala) [reply socket-ballerina/build.gradle:88]: This is relevant to the `bir-cache` directory structure 
```
jballerina-1.3.1-SNAPSHOT/bir-cache/ballerina/socket/0.0.0/socket.bir
```

File: socket-ballerina/build.gradle (hunk 6 above)
Comment (BuddhiWathsala) [reply socket-ballerina/build.gradle:36]: Fixed in https://github.com/ballerina-platform/module-ballerina-socket/pull/2/commits/99fccd29af679a7560130308c6579f021f53c1cf

File: gradle/wrapper/gradle-wrapper.properties (hunk 5 above)
Comment (BuddhiWathsala) [reply gradle/wrapper/gradle-wrapper.properties]: Fixed in https://github.com/ballerina-platform/module-ballerina-socket/pull/2/commits/7830013d54028367951e16e6456910ec8bba1be5

**Pull request merged**